### PR TITLE
[RFC, RFT] realtek: add Kernel 5.10 support

### PR DESCRIPTION
**Pull request by musashino205**
This patch series adds Kernel 5.10 support to realtek target.

summary:

- platform support: based on current master (5.4)
- RTL8231 support: use exsisting driver (5.4)
- backported drivers from upstream:
  - RTL83xx GPIO controller (gpio-realtek-otto, 5.13)
  - IRQ controller (irq-realtek-rtl, 5.12)
  - SPI controller (spi-realtek-rtl, 5.12)
- add "dts-5.10" directory

Tested on:

- INABA Abaniact AML2-17GP (RTL8382M, 17ports)
- I-O DATA BSH-G24MB (RTL8382M, 24ports, WIP device)
- Panasonic Switch-M8eG PN28080K (RTL8380M, 8ports + 1 SFP, WIP device)

### CODE REVIEW

File: target/linux/realtek/dts-5.10/rtl8382_allnet_all-sg8208m.dts [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+	keys {
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
+
+		/* is this pin 3 on the external RTL8231 (&gpio1)? */
+		/*reset {
+			label = "reset";
+			gpios = <&gpio0 67 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};*/
```
Comment (musashino205) [inline target/linux/realtek/dts-5.10/rtl8382_allnet_all-sg8208m.dts:37]: Q: ```&gpio0 67``` -> ```&gpio1 3``` ?

File: target/linux/realtek/dts-5.10/rtl8382_d-link_dgs-1210-10p.dts [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+	keys {
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
+
+		/* is this pin 30 on the external RTL8231 (&gpio1)? */
+		/*mode {
+			label = "reset";
+			gpios = <&gpio0 94 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};*/
```
Comment (musashino205) [inline target/linux/realtek/dts-5.10/rtl8382_d-link_dgs-1210-10p.dts:49]: Q: ```&gpio0 94``` -> ```&gpio1 30``` ?

**Comment by dangowrt** [comment]
Tested on ZyXEL GS1900-10HP (SFP ports untested)
[gs1900hp-boot.txt](https://github.com/openwrt/openwrt/files/6987321/gs1900hp-boot.txt)

**Comment by musashino205** [comment]
tested SFP on Panasonic Switch-M8eG PN28080K (RTL8380M, 8x TP + 1x SFP, WIP device):

- SFP module: 2x Cisco GLC-SX-MM
- opposite device: Panasonic Switch-M24eG PN28240K (stock)

https://gist.github.com/musashino205/3a60d7a024effe13dfe4d688046278ac

**Comment by bmork** [comment]
Tested on Netgear GS108Tv3.  Nice work!
[gs108tv3-boot.txt](https://github.com/openwrt/openwrt/files/6990301/gs108tv3-boot.txt)

**Comment by bkobl** [comment]
Great work!
I am having problems with the 930x devices, though. The GPIO's dont work. Booting hangs at

`[    0.000000] Linux version 5.10.56 (birger@ThinkPadAMD) (mips-openwrt-linux-musl-gcc (OpenWrt GCC 8.4.0 r17321+31-5c88bf896e) 8.4.0, GNU ld (GNU Binutils) 2.34) #0 Fri Aug 13 15:27:14 2021
[    0.000000] RTL838X model is 0
[    0.000000] RTL839X model is 0
[    0.000000] RTL93XX model is 93021001
[    0.000000] SoC Type: RTL9302B
[    0.000000] Kernel command line: 
[    0.000000] printk: bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019555 (MIPS 34Kc)
[    0.000000] MIPS: machine is Zyxel XGS1210-12 Switch
[    0.000000] Registering _machine_restart
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Using appended Device Tree.
[    0.000000] Primary instruction cache 32kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] On node 0 totalpages: 32768
[    0.000000]   Normal zone: 288 pages used for memmap
[    0.000000]   Normal zone: 0 pages reserved
[    0.000000]   Normal zone: 32768 pages, LIFO batch:7
[    0.000000] pcpu-alloc: s0 r0 d32768 u32768 alloc=1*32768
[    0.000000] pcpu-alloc: [0] 0 
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 32480
[    0.000000] Kernel command line: console=ttyS0,115200
[    0.000000] Dentry cache hash table entries: 16384 (order: 4, 65536 bytes, linear)
[    0.000000] Inode-cache hash table entries: 8192 (order: 3, 32768 bytes, linear)
[    0.000000] Writing ErrCtl register=00060163
[    0.000000] Readback ErrCtl register=00060163
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 115024K/131072K available (5330K kernel code, 608K rwdata, 1180K rodata, 7400K init, 203K bss, 16048K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 256
[    0.000000] random: get_random_bytes called from start_kernel+0x3c4/0x5b4 with crng_init=0
[    0.000000] rtl9300_timer_init: setting up timer
[    0.000000] Frequency in dts: 175000000, my rate is 87500000, period 1750000
[    0.000000] With base b8003200 IRQ: 8
[    0.000000] clocksource: timer: mask: 0xfffffff max_cycles: 0xfffffff, max_idle_ns: 1365186028 ns
[    0.000008] sched_clock: 28 bits at 87MHz, resolution 11ns, wraps every 1533916858ns
[    0.008454] CPU frequency from device tree: 800MHz
[    0.008481] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 4778151116 ns
[    0.023354] sched_clock: 32 bits at 400MHz, resolution 2ns, wraps every 5368709118ns
[    0.031840] Calibrating delay loop... 1064.96 BogoMIPS (lpj=5324800)
[    0.115328] pid_max: default: 32768 minimum: 301
[    0.120499] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.128362] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.138852] dyndbg: Ignore empty _ddebug table in a CONFIG_DYNAMIC_DEBUG_CORE build
[    0.149438] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.160058] futex hash table entries: 256 (order: -1, 3072 bytes, linear)
[    0.167558] pinctrl core: initialized pinctrl subsystem
[    0.173777] NET: Registered protocol family 16
[    0.223219] clocksource: Switched to clocksource MIPS
[    0.229840] NET: Registered protocol family 2
[    0.234661] IP idents hash table entries: 2048 (order: 2, 16384 bytes, linear)
[    0.242986] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.252112] TCP established hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.260363] TCP bind hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.268020] TCP: Hash tables configured (established 1024 bind 1024)
[    0.275097] UDP hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.282135] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.290029] NET: Registered protocol family 1
[    0.432815] workingset: timestamp_bits=14 max_order=15 bucket_order=1
[    0.443793] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.450131] jffs2: version 2.2 (NAND) (SUMMARY) (ZLIB) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
`

Kernel 5.4 would continue like this:
`[    0.490660] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.496912] jffs2: version 2.2 (NAND) (SUMMARY) (ZLIB) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.527500] Probing RTL838X GPIOs for 00009302
[    0.532295] Found RTL9302 GPIO
`
I dont know whether this is a dts problem (am using the one provided) or a problem of the driver. I am not sure that driver was developed for the newer 93xx SoCs.

**Comment by svanheule** [comment]
>  am having problems with the 930x devices, though. The GPIO's dont work.

That's to be expected. The port order is reversed compared to rtl83xx. Since I didn't have any 93xx hardware to test, I didn't implement those GPIO peripherals yet.

As a stop-gap solution, I think you should be able to use the fallback compatible with 32 pins, and reversed port order.

**Comment by musashino205** [comment]
Thank you!
I didn't know details about the new driver with RTL93xx SoCs too, but there is a SoC gpio node with ```compatible = "realtek,rtl838x-gpio"``` in rtl930x.dtsi and I guessed the driver can also be used for RTL930x.dtsi. (I don't have any device with RTL930x SoC...)

> That's to be expected. The port order is reversed compared to rtl83xx. Since I didn't have any 93xx hardware to test, I didn't implement those GPIO peripherals yet.

Oh, I see. There is no officially supported RTL930x device in OpenWrt yet, so I would like to temporarily disable or remove the SoC GPIO node in rtl930x.dtsi for 5.10.

**Comment by bkobl** [comment]
Disabling the 93xx GPIO support for now is fine for me. It is more important that the backported drivers plus the 5.10 support make it into OpenWRT and we can build up from there.

**Comment by musashino205** [comment]
UPDATE:

- disable SoC GPIO node (```gpio0```) in rtl930x.dtsi (9dbd5278b789d6002edc94d010f6c635d63aa694)
- drop ```indirect-access-bus-id``` property from device dts files in (9ad67df7ff1c98ebd74e27887206c7020e5d578e)

**Comment by Neustradamus** [comment]
To follow this PR

**Comment by Borromini** [comment]
I just tested this on my GS1900-8HP v1. Seems to works nice, kudos @musashino205 

Question: PoE doesn't seem to turn off between reboots, is that considered normal (or desired) behaviour? I know both my GS1900-10HP and the -8HP v1 would kill PoE clients when rebooting (same with early support in master and now in the 21.02 branch).

I tested this alongside blogic's PoE C implementation, which doesn't seem to pick up anything of the PoE internals . Still have to reference that with the Lua implementation on 5.10 though.

**Comment by bmork** [comment]
> Question: PoE doesn't seem to turn off between reboots, is that considered normal (or desired) behaviour?

Interesting!  I'd definitely say it's desired.

**Comment by musashino205** [comment]
Thank you for the testing, I have no device with PoE support and didn't test about it...
The micro-controller for PoE is controlled via secondary UART on SoC? If so, I've disabled it in SoC dtsi files and it need to be enabled in device dts to use.

**Comment by musashino205** [comment]
UPDATE:

- enable ```uart1``` on the following devices for PoE
  - D-Link DGS-1210-10P
  - Netgear GS110TPP v1
  - Netgear GS310TP v1
  - Zyxel GS1900-8HP v1/v2
  - Zyxel GS1900-10HP

**Comment by Borromini** [comment]
@musashino205 Thanks, that makes the PoE daemon output valid info indeed!

```
# ubus call poe info
{
	"firmware": "v17.1",
	"mcu": "ST Micro ST32F100 Microcontroller",
	"budget": 70.000000,
	"consumption": 3.100000,
	"ports": {
		"lan1": {
			"priority": 2,
			"mode": "PoE+",
			"status": "Searching"
		},
		"lan2": {
			"priority": 2,
			"mode": "PoE+",
			"status": "Delivering power",
			"consumption": 3.200000
		}
	}
}
```

あろがとう！

**Review by dangowrt — APPROVED** [review]
Run-tested on ZyXEL GS1900-10HP, uptime of 10 days for now, all good.
I've looked at all the commits and didn't see any issues.

**Comment by Neustradamus** [comment]
What is the blocking part before a merging?

**Comment by musashino205** [comment]
UPDATE:

- rebase from master for GS1900-24HP v2
  - enable uart1 for GS1900-24HP v2

**Comment by adschm** [comment]
Since this already contains quite a lot of patches already and has received positive feedback, I consider merging it during the weekend so it's easier to go on with it (it's just testing for now).

I've reordered the patches a bit, moved dts to dts-5.4 and am currently doing build-testing ...

**Comment by adschm** [comment]
I've merged it. Thanks for bringing everything into a nice shape.

**Pull request closed without merging**